File: houses/submission1.cpp
/*
    Language: C++20
    Submitted By: 
        Umesh Yadav 
        (github.com/umesh72614)
        (hackerrank.com/2018ucs0078)
    Score: 10/10
*/ 

#include <vector>
#include <iostream>
#include <algorithm>
using namespace std;

void dfs(vector<vector<int>> &adj, vector<bool> &vis, int i) {
    cout << i << "\n";
    vis[i] = true;
    for (auto &j: adj[i])
        if (!vis[j]) dfs(adj, vis, j);
}

int main() {
    int n;
    cin >> n;
    vector<vector<int>> adj(n);
    vector<int> indeg(n);
    for (int i = 0; i < n; i++) {
        int x, y;
        cin >> x >> y;
        adj[x].push_back(y);
        indeg[y]++;
    }
    vector<bool> vis(n);
    for (int i = 0; i < n; i++)
        if (!indeg[i]) dfs(adj, vis, i);
}

File: maximize_integer/submission1.py
"""
    Language: python3
    Submitted By: 
        Umesh Yadav 
        (github.com/umesh72614)
        (hackerrank.com/2018ucs0078)
    Score: 40/40
"""

# use python as x can be very very large
# also no need to convert x into int as <,> are valid in strings also
x = input()
rx = x[::-1]
n = len(x)  # num of digits
d = max(list(x))    # maxDigit
m = '0'     # store ans
for i in range(1, n):
    # don't perform operation if digit is not maxDigit (GREEDY)
    if (x[i-1] != d):
        continue
    # y = x[:i][::-1] + x[i:][::-1]
    # Below is the optimization to avoid reverse at each step
    y = rx[n-i:] + rx[:n-i]
    # if y is smaller than m then discard it
    if y > m:
        m = y
print(x if m == '0' else m)
